File: server/services/shoppingService.js
/**
 * 购物清单服务
 */
class ShoppingService {
  /**
   * 生成购物清单
   */
  async generateList({ dishes, servings, existingIngredients }) {
    // 模拟食材数据库（实际项目中应该从数据库或AI获取）
    const ingredientDatabase = {
      '红烧肉': [
        { name: '五花肉', amount: '500g', category: '肉类' },
        { name: '生姜', amount: '30g', category: '调料' },
        { name: '大葱', amount: '2根', category: '蔬菜' },
        { name: '八角', amount: '3个', category: '调料' },
        { name: '老抽', amount: '2勺', category: '调料' },
        { name: '冰糖', amount: '50g', category: '调料' }
      ],
      '清炒时蔬': [
        { name: '时令蔬菜', amount: '300g', category: '蔬菜' },
        { name: '大蒜', amount: '3瓣', category: '调料' },
        { name: '盐', amount: '适量', category: '调料' }
      ],
      '番茄蛋汤': [
        { name: '番茄', amount: '2个', category: '蔬菜' },
        { name: '鸡蛋', amount: '2个', category: '蛋类' },
        { name: '香油', amount: '几滴', category: '调料' }
      ]
    };

    // 收集所有需要的食材
    let allIngredients = [];
    dishes.forEach(dish => {
      const ingredients = ingredientDatabase[dish.name] || [];
      // 根据人数调整分量
      const adjustedIngredients = ingredients.map(ing => ({
        ...ing,
        amount: this.adjustAmount(ing.amount, servings)
      }));
      allIngredients = allIngredients.concat(adjustedIngredients);
    });

    // 合并相同食材
    const mergedIngredients = this.mergeIngredients(allIngredients);

    // 标记已有食材
    const finalList = mergedIngredients.map(item => ({
      ...item,
      status: existingIngredients.includes(item.name) ? '已有' : '需购买',
      purchaseTip: this.getPurchaseTip(item)
    }));

    // 按分类排序
    const categorizedList = this.categorizeIngredients(finalList);

    return {
      summary: {
        totalItems: finalList.length,
        needToBuy: finalList.filter(i => i.status === '需购买').length,
        alreadyHave: finalList.filter(i => i.status === '已有').length
      },
      ingredients: categorizedList
    };
  }

  /**
   * 调整食材分量
   */
  adjustAmount(amount, servings) {
    // 简单的分量调整逻辑
    const baseServings = 2;
    if (servings === baseServings) return amount;
    
    const multiplier = servings / baseServings;
    const match = amount.match(/(\d+)(.*)/);
    
    if (match) {
      const num = parseInt(match[1]);
      const unit = match[2];
      return `${Math.round(num * multiplier)}${unit}`;
    }
    
    return amount;
  }

  /**
   * 合并相同食材
   */
  mergeIngredients(ingredients) {
    const merged = {};
    
    ingredients.forEach(item => {
      if (merged[item.name]) {
        // 简单的合并逻辑（实际项目需要更复杂的单位换算）
        merged[item.name].amount += ` + ${item.amount}`;
      } else {
        merged[item.name] = { ...item };
      }
    });
    
    return Object.values(merged);
  }

  /**
   * 获取购买建议
   */
  getPurchaseTip(ingredient) {
    const tips = {
      '肉类': '建议选择新鲜、有光泽的肉品',
      '蔬菜': '选择叶片翠绿、无黄叶的新鲜蔬菜',
      '调料': '可在超市一次性购齐常用调料',
      '蛋类': '检查蛋壳完整，无裂纹'
    };
    
    return tips[ingredient.category] || '根据个人喜好选择';
  }

  /**
   * 按分类组织食材
   */
  categorizeIngredients(ingredients) {
    const categories = {};
    
    ingredients.forEach(item => {
      if (!categories[item.category]) {
        categories[item.category] = [];
      }
      categories[item.category].push(item);
    });
    
    return categories;
  }

  /**
   * 更新库存
   */
  async updateInventory(ingredients) {
    // 实际项目中应该保存到数据库
    return {
      updated: true,
      count: ingredients.length,
      timestamp: new Date().toISOString()
    };
  }
}

module.exports = new ShoppingService();

File: server-rust/src/services/shopping_service.rs
use crate::error::AppError;
use crate::models::{Ingredient, ShoppingList, ShoppingSummary};
use std::collections::HashMap;

pub struct ShoppingService;

impl ShoppingService {
    pub fn new() -> Self {
        Self
    }

    pub async fn generate_list(
        &self,
        dishes: Vec<String>,
        servings: u8,
        existing_ingredients: Vec<String>,
    ) -> Result<ShoppingList, AppError> {
        // 模拟食材数据库
        let ingredient_db = self.get_ingredient_database();

        // 收集所有食材
        let mut all_ingredients: Vec<Ingredient> = Vec::new();

        for dish_name in dishes {
            if let Some(ingredients) = ingredient_db.get(&dish_name) {
                let adjusted = ingredients
                    .iter()
                    .map(|ing| self.adjust_amount(ing.clone(), servings))
                    .collect::<Vec<_>>();
                all_ingredients.extend(adjusted);
            }
        }

        // 合并相同食材
        let merged = self.merge_ingredients(all_ingredients);

        // 标记已有食材
        let final_list = merged
            .into_iter()
            .map(|mut ing| {
                let status = if existing_ingredients.contains(&ing.name) {
                    "已有".to_string()
                } else {
                    "需购买".to_string()
                };
                ing.status = Some(status);
                ing.purchase_tip = Some(self.get_purchase_tip(&ing.category));
                ing.purchased = Some(false);
                ing
            })
            .collect();

        // 分类
        let categorized = self.categorize_ingredients(final_list);

        // 统计
        let total_items: usize = categorized.values().map(|v| v.len()).sum();
        let need_to_buy = categorized
            .values()
            .flat_map(|v| v.iter())
            .filter(|i| i.status.as_deref() == Some("需购买"))
            .count();
        let already_have = total_items - need_to_buy;

        Ok(ShoppingList {
            summary: ShoppingSummary {
                total_items,
                need_to_buy,
                already_have,
            },
            ingredients: categorized,
        })
    }

    fn get_ingredient_database(&self) -> HashMap<String, Vec<Ingredient>> {
        let mut db = HashMap::new();

        db.insert(
            "红烧肉".to_string(),
            vec![
                Ingredient {
                    name: "五花肉".to_string(),
                    amount: "500g".to_string(),
                    category: "肉类".to_string(),
                    status: None,
                    purchase_tip: None,
                    purchased: None,
                },
                Ingredient {
                    name: "生姜".to_string(),
                    amount: "30g".to_string(),
                    category: "调料".to_string(),
                    status: None,
                    purchase_tip: None,
                    purchased: None,
                },
                Ingredient {
                    name: "大葱".to_string(),
                    amount: "2根".to_string(),
                    category: "蔬菜".to_string(),
                    status: None,
                    purchase_tip: None,
                    purchased: None,
                },
                Ingredient {
                    name: "八角".to_string(),
                    amount: "3个".to_string(),
                    category: "调料".to_string(),
                    status: None,
                    purchase_tip: None,
                    purchased: None,
                },
            ],
        );

        db.insert(
            "清炒时蔬".to_string(),
            vec![
                Ingredient {
                    name: "时令蔬菜".to_string(),
                    amount: "300g".to_string(),
                    category: "蔬菜".to_string(),
                    status: None,
                    purchase_tip: None,
                    purchased: None,
                },
                Ingredient {
                    name: "大蒜".to_string(),
                    amount: "3瓣".to_string(),
                    category: "调料".to_string(),
                    status: None,
                    purchase_tip: None,
                    purchased: None,
                },
            ],
        );

        db.insert(
            "番茄蛋汤".to_string(),
            vec![
                Ingredient {
                    name: "番茄".to_string(),
                    amount: "2个".to_string(),
                    category: "蔬菜".to_string(),
                    status: None,
                    purchase_tip: None,
                    purchased: None,
                },
                Ingredient {
                    name: "鸡蛋".to_string(),
                    amount: "2个".to_string(),
                    category: "蛋类".to_string(),
                    status: None,
                    purchase_tip: None,
                    purchased: None,
                },
            ],
        );

        db
    }

    fn adjust_amount(&self, mut ingredient: Ingredient, servings: u8) -> Ingredient {
        let base_servings = 2u8;
        if servings == base_servings {
            return ingredient;
        }

        let multiplier = servings as f32 / base_servings as f32;

        // 简单的分量调整
        if let Some(num) = ingredient.amount.chars().next() {
            if num.is_numeric() {
                if let Some(pos) = ingredient.amount.find(|c: char| !c.is_numeric()) {
                    if let Ok(original_num) = ingredient.amount[..pos].parse::<u32>() {
                        let adjusted = (original_num as f32 * multiplier) as u32;
                        ingredient.amount = format!("{}{}", adjusted, &ingredient.amount[pos..]);
                    }
                }
            }
        }

        ingredient
    }

    fn merge_ingredients(&self, ingredients: Vec<Ingredient>) -> Vec<Ingredient> {
        let mut merged: HashMap<String, Ingredient> = HashMap::new();

        for ing in ingredients {
            let key = ing.name.clone();
            if let Some(existing) = merged.get_mut(&key) {
                // 简单合并：保留第一个的量
                // 实际项目中需要更复杂的单位换算
            } else {
                merged.insert(key, ing);
            }
        }

        merged.into_values().collect()
    }

    fn get_purchase_tip(&self, category: &str) -> String {
        match category {
            "肉类" => "建议选择新鲜、有光泽的肉品".to_string(),
            "蔬菜" => "选择叶片翠绿、无黄叶的新鲜蔬菜".to_string(),
            "调料" => "可在超市一次性购齐常用调料".to_string(),
            "蛋类" => "检查蛋壳完整，无裂纹".to_string(),
            _ => "根据个人喜好选择".to_string(),
        }
    }

    fn categorize_ingredients(&self, ingredients: Vec<Ingredient>) -> HashMap<String, Vec<Ingredient>> {
        let mut categorized: HashMap<String, Vec<Ingredient>> = HashMap::new();

        for ing in ingredients {
            categorized
                .entry(ing.category.clone())
                .or_insert_with(Vec::new)
                .push(ing);
        }

        categorized
    }
}
